### PR TITLE
feat: ERPNext support for Postgres

### LINKED ISSUE #24389: Postgres support for ERPNext
Frappe Framework has support for both MariaDB and Postgres. ERPNext does not have support for Postgres. While queries are mostly similar, there are 1300 SQL queries that need to be verified and checked. Thankfully we have unit-tests.

Steps:

1. Setup tests for postgres (similar to Frappe Framework)
2. Makes sure the tests pass for all queries

If there are problems in certain SQL queries:

1. Use the ORM
2. Rewrite the query so that it will work for both databases
3. Use `frappe.db.multi_sql` to write separate queries for each platform (for complex joins)

### PR DESCRIPTION
**Pull request by c-p-b**
Companion Frappe PR: https://github.com/frappe/frappe/pull/15354

At a high level, mostly multisql was used when possible favoring changing MariaDB queries themselves. When the query had to be rewritten entirely, db independent syntax and frappe.db.sql was favored. The goal of this PR is to target as low-risk change as possible for existing installations.  It adds a bit of technical debt in that in the ideal world, ORM would be used where possible, there are perhaps several simpler queries that I touched that could have converted to using the ORM, but at the risk of introducing new bugs/behavior in the MariaDB queries I opted not to do that and instead favored using multisql and frappe.db.sql.

Caveats:

1. Little was known about underlying data behavior or what things are used for - As a result I tried to change the existing behavior of mdb queries as little as possible. Because of this, ORM was only minimally used, and generally the strategy of modifying existing query syntax to support both engines where possible and using multisql where not possible was followed.  If conversion to multisql became too ugly/unweildy, extra time was taken to port to the ORM
2. This was only tested against Postgres 13.2 and MariaDB 10.3
3. During pg date/time conversions and comparisons some explicit type casting was required for the engine to execute functions - some of this may  throw different errors in runtime than what you might see from mariadb and this also may introduce some performance implications
4. Using the collation for MariaDB that is by default provided by ERPNext means that queries are case insensitive in MariaDB. This PR accounts for that and changes reflect that queries are case sensitive for Postgres for the purposes of better data integrity. Therefore Postgres is more stringent than MariaDB in this regard
5. Currently this PR has changes in frappe that require auto_commit on certain calls to db_insert() and insert() functions. These auto_commits only run during installation and testing and only in specific parts of test data bootstrap. During normal runtime normal behavior is respected
6. The data structure of the output of several queries needed to be modified. This primarily affects conversions from frappe.db.get_* to frappe.qb and frappe.db.sql/multisql. Typically the old style queries would return nested tuples e.g. ((, )), and the newer style returns a tuple wrapped in a list e.g. [(, )]. The output of the queries is pretty well covered in tests as they are all consumed downstream so this should be a relatively low risk change.

In this PR:

1. Fixed all instances of the sql directly checking against the timestamp ‘0000-00…’ as this is not a valid date in Postgres (but it is for MariaDB). It is instead removed entirely as the Postgres equivalent is simply NULL which the MariaDB queries already check for
2. Converted all instances calling timestamp() to its equivalent addition operator (date + time) in pgsql
3. Changed all instances of unix_timestamp() to extract(epoch from)
4. Removed ‘force index’ hint from pgsql queries as Postgres does not support it 
5. All instances of time_format()/date_format() converted to Postgres equivalent to_char
6. Any instances of LIMIT 1,2 updated to LIMIT 2 OFFSET 1 which is valid syntax in both Postgres and MariaDB
7. Changed ifnull(x) -> is null or coalesce() depending on usage and check variable empty strings by casting to text first
8. Changed any unquoted table/column names starting with lowercase to ensure they are wrapped in `backtick` which Postgres requires
9. Removed ambiguity of certain clauses by explicitly specifying column. Specifically, in MariaDB, you are not required to specify the column names from the SELECT clause in the GROUP BY clause. In Postgres, you are required to specify these column names. 
10. MariaDB allows either double quote or single quote for value lookup in query. Postgres only allow single quote. Converted any explicit double quoting to single quote in query
11. MariaDB IF statement converted to CASE WHEN THEN ELSE END
12. Removed ORDER BY NULL as it no longer impacts query performance: https://dev.mysql.com/doc/refman/8.0/en/order-by-optimization.html
13. Replaced CURDATE() instances with CURRENT_DATE which works in both MariaDB and Postgresql 
14. Date_Sub function in MariaDB changed to subtraction operand for Postgresql
15. Updated tests in instances where case sensitivity and minor data structure changes were required

TODO:

- [x] Resolve open merge conflicts
- [x] CI should run tests using Postgres as well in ERPNext - perhaps someone from ERPNext can assist here as I’m not privy to the details of how you will like your CI setup to work with multiple databases in ERPNext.
- [x] Verify/fix failing tests for MariaDB and Postgres
- [x] TBD with more detail - Break out changes into individual PR's for review and mainlining

### CODE REVIEW

**Comment by c-p-b** [comment]
These are the steps I used to build and test locally:

```bash
 bench init test-frappe-dev --frappe-path=https://github.com/cpdeethree/frappe.git
 bench new-site test-erp-postgres --db-host my-pg-host-here --db-port 5432 --db-type postgres
 bench new-site test-erp-mdb --db-host my-mdb-host-here
 bench get-app erpnext git@github.com:cpdeethree/erpnext.git 
 bench --site test-erp-postgres install-app erpnext
 bench --site test-erp-mdb install-app erpnext
 bench config dns_multitenant off
 rm sites/currentsite.txt
```
 
 add following to /etc/hosts

 127.0.0.1        test_frappe
 127.0.0.1        test_api
 127.0.1.1        myusername

 ```bash
 bench setup supervisor
 bench restart
 bench start --site whichever-site-we-are-testing
 Navigate to localhost:8000 and fill out the information to register the site in English. Select India as country. Use ‘example’ or ‘e’ or ‘example.com’ for all fields
 bench --site test-erp-postgres set-config allow_tests true
 bench --site test-erp-mdb set-config allow_tests true
 bench run-tests
```

**Comment by c-p-b** [comment]
@rmehta I've gotten most of the low hanging fruit postgres test failures fixed. Can we get some assistance in getting CI to target postgres for tests as well? That should also help narrow down and clean up the remaining postgres failing tests as I'm not sure which ones that were failing on my machine were due to dependency or setup related issues. Currently still working on cleaning up the MariaDB test failures from the changes as well.

**Comment by rmehta** [comment]
Thanks @cpdeethree for this!

We would love to collaborate. Let's get on the contributors TG group to take it forward: https://t.me/erpnextcontrib.

Love if someone from Frappe takes ownership of this contribution @nabinhait @ankush - can you guys figure this out.

The first target would be to get the tests passed!

File: erpnext/controllers/item_variant.py [hunk 1]
```diff
@@ -309,13 +309,19 @@ def make_variant_item_code(template_item_code, template_item_name, variant):
...
 	for attr in variant.attributes:
-		item_attribute = frappe.db.sql("""select i.numeric_values, v.abbr
+		item_attribute = frappe.db.multisql({
+			'mariadb':"""select i.numeric_values, v.abbr
 			from `tabItem Attribute` i left join `tabItem Attribute Value` v
 				on (i.name=v.parent)
-			where i.name=%(attribute)s and (v.attribute_value=%(attribute_value)s or i.numeric_values = 1)""", {
+			where i.name=%(attribute)s and (v.attribute_value=%(attribute_value)s or i.numeric_values = 1)""",
+			'postgres': """select i.numeric_values, v.abbr
+			from `tabItem Attribute` i left join `tabItem Attribute Value` v
```
Comment (klando) [inline erpnext/controllers/item_variant.py:316]: @cpdeethree note that backstick ` is not legal SQL. It's double quote around tables, columns, .. and simple quote for values.

**Comment by klando** [comment]
@cpdeethree looks like you're very active on this PR.
What's the current status ? I would really like to use ERPnext with PostgreSQL backend and might be able to help. What's the quick and easy way to test this PR (docker compose maybe) ?

**Comment by klando** [comment]
@cpdeethree it might be easier to merge by splitting the PR in 2: one to add multisql (hopefully no regression test changes required) and one to add PostgreSQL, more complex...and maybe easier to review with a more limited scope.

**Comment by klando** [comment]
> @cpdeethree it might be easier to merge by splitting the PR in 2: one to add multisql (hopefully no regression test changes required) and one to add PostgreSQL, more complex...and maybe easier to review with a more limited scope.

and still in order to reduce patch size related to PostgreSQL, given that MySQL supports `COALESCE()` would it be possible to replace the MySQLish `IFNULL()`  all over the place ?

**Comment by klando** [comment]
it looks like some tests are failing due to lines order. Don't know for MySQL but with PostgreSQL you need to add an `ORDER BY` to make sure of that.

**Comment by c-p-b** [comment]
@klando Current status is noted as above in the checkmarks - but in general most of the initial work is done (down to a few remaining test failures from over 700!) but there are a couple of things still needing testing/fixes. If you are interested in collaborating, please join the telegram channel @rmehta mentioned above and make a comment there - I will get in touch with you there

File: erpnext/controllers/item_variant.py (hunk 1 above)
Comment (c-p-b) [reply erpnext/controllers/item_variant.py:316]: Frappe under the hood substitutes this to legal sql

**Comment by vpicavet** [comment]
Great ! PostgreSQL support would be a game-changer for us !

**Comment by ankush** [comment]
I am unable to focus on this and right now ERPNext team is rewriting many queries to QB which will eventually make it simpler to run it on postgres. 

ref: 
- https://github.com/frappe/erpnext/pull/32520
- https://github.com/frappe/erpnext/pull/32527
- many more in past

This will take time but it's on the right course. We need to ensure maintainability and this is the right approach to handle multiple DBs 😅 

Thanks a lot for the efforts, we have merged quite a lot of it separately and more will eventually get merged piece by piece (we are still using this PR as reference 😄 ) It just wasn't ready for a one-shot merge. 

Closing for now.

**Comment by infogulch** [comment]
Is there a tracking issue for the goal to support multiple databases?

**Comment by v3ss0n** [comment]
What happened to the postgres support ? how well ERPNext support it now?

**Comment by gowthamHutech** [comment]
can you just post the commands to connect with postgres from erp next


gowtham@gowtham-VirtualBox:~/frappe-bench$ bench new-site google.com --db-type postgres
Postgres super user password: 
Restoring Database file...
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:9: NOTICE:  table "tabDocField" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:72: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDocField" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:74: ERROR:  relation "tabDocField" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:75: ERROR:  relation "tabDocField" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:76: ERROR:  relation "tabDocField" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:77: ERROR:  relation "tabDocField" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:83: NOTICE:  table "tabDocPerm" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:112: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDocPerm" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:114: ERROR:  relation "tabDocPerm" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:120: NOTICE:  table "tabDocType Action" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:137: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDocType Action" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:139: ERROR:  relation "tabDocType Action" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:145: NOTICE:  table "tabDocType Link" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:162: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDocType Link" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:164: ERROR:  relation "tabDocType Link" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:171: NOTICE:  table "tabDocType" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:236: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDocType" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:242: NOTICE:  table "tabSeries" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:247: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabSeries" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:253: NOTICE:  table "tabSessions" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:262: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabSessions" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:264: ERROR:  relation "tabSessions" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:270: NOTICE:  table "tabSingles" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:275: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabSingles" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:277: ERROR:  relation "tabSingles" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:283: NOTICE:  table "__Auth" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:291: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "__Auth" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:293: ERROR:  relation "__Auth" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:299: NOTICE:  table "tabFile" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:318: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabFile" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:320: ERROR:  relation "tabFile" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:321: ERROR:  relation "tabFile" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:322: ERROR:  relation "tabFile" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:328: NOTICE:  table "tabDefaultValue" does not exist, skipping
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:343: ERROR:  permission denied for schema public
LINE 1: CREATE TABLE "tabDefaultValue" (
                     ^
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:345: ERROR:  relation "tabDefaultValue" does not exist
psql:/home/gowtham/frappe-bench/apps/frappe/frappe/database/postgres/framework_postgres.sql:346: ERROR:  relation "tabDefaultValue" does not exist
Table 'tabDefaultValue' missing in the restored site. This may be due to incorrect permissions or the result of a restore from a bad backup file. Database not installed correctly.


how can i resolve these

**Comment by edardev** [comment]
Does ERPNext support Postgres?? I tried using docker compose and it does not work...

**Pull request closed without merging**